File: python/tn/handlers/firehose.py
"""TN firehose handler: encrypted log streaming to the TN vault.

Streams encrypted log frames over a long-lived WebSocket to the TN
vault, which forwards (via service-binding RPC) to the CF firehose
Worker, which buffers and writes batched `.tnpkg` artifacts to R2.

Zero-knowledge: each frame is encrypted client-side under the project
BEK (looked up by ``key_id`` in the local keystore) before it leaves
the device. The vault and CF Worker only ever see ciphertext.

Phase A status: this is the handler SCAFFOLD. The following pieces are
wired:

* Class structure (inherits ``AsyncHandler`` for outbox + worker drain).
* Config schema (endpoint, project_id, key_id, filter_spec).
* Filter spec support (inherited from base).
* ``_publish`` stub that validates inputs and records what would have
  been sent. Real WebSocket connection + reconnect logic ships in B3.

The following pieces are intentionally stubbed and ship later in Phase A:

* B3: real WebSocket client (open, reconnect-with-backoff, send).
* B4: frame encryption with project BEK from keystore. Phase A uses a
  test-stub BEK source so the handler can be exercised end-to-end
  before Phase B's manifest schema changes land.

Config shape (``tn.yaml``)::

    handlers:
      - kind: tn.firehose
        name: cloud-archive
        endpoint: https://vault.tn-proto.org
        project_id: 00000000-0000-0000-0000-000000000000
        key_id: fhk_<base32>             # optional in Phase A (test stub)
        filter:
          level_in: [info, warning, error]

Spec reference: ``docs/superpowers/specs/2026-05-15-firehose-design.md``
``§ Client handler (tn.firehose)``.
"""

from __future__ import annotations

import hashlib
import logging
import os
import re
import threading
from typing import Any
from urllib.parse import urlparse, urlunparse

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from .base import AsyncHandler

_log = logging.getLogger("tn.handlers.firehose")


def _stub_bek(project_id: str, key_id: str | None) -> bytes:
    """Phase A test-stub BEK derivation.

    Derives a deterministic 32-byte AES key from ``(project_id,
    key_id)`` so the encryption path can be exercised end-to-end before
    Phase B's real keystore integration lands.

    **This is not secure.** Two devices with the same project_id and
    key_id derive the same key, which is the OPPOSITE of what the real
    sealed-box flow does (only authorized member devices hold the BEK).
    Phase B replaces this function with ``keystore.lookup_bek(key_id)``
    which returns bytes provisioned via the mint flow.
    """
    seed = f"{project_id}:{key_id or 'stub-default'}".encode("utf-8")
    return hashlib.sha256(b"phase-a-stub-bek-do-not-use-in-prod:" + seed).digest()

# Strict UUID v4-shaped project id (matches the CF Worker's UUID_RE).
_PROJECT_ID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)

# Loose key_id format; production keys are minted by the vault UI and
# follow a "fhk_<base32>" shape, but Phase A accepts anything for
# test-stub purposes.
_KEY_ID_RE = re.compile(r"^[A-Za-z0-9_-]{1,128}$")


class TnFirehoseHandler(AsyncHandler):
    """Streaming firehose handler.

    Constructor parameters
    ----------------------
    name
        Handler instance name. Used in logs and outbox path.
    endpoint
        Base URL of the TN vault, e.g. ``https://vault.tn-proto.org``.
        The vault is the only network endpoint the handler talks to;
        the CF Worker behind it is invisible to clients.
    project_id
        UUID identifying the project this handler streams under. Must
        match a project record at the vault that has the bearer DID as
        a member.
    key_id
        Optional. Identifier of the active project BEK to encrypt
        frames with. In Phase A this is a test-stub identifier; B4
        wires real BEK lookup from the local keystore.
    filter_spec
        Optional standard handler filter dict (``level_in``,
        ``event_type_prefix``, etc.). Pre-filtered envelopes don't
        reach ``_publish`` so the WS doesn't carry events the user
        doesn't want shipped.
    """

    def __init__(
        self,
        name: str,
        outbox_path: Any,
        *,
        endpoint: str,
        project_id: str,
        key_id: str | None = None,
        filter_spec: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(name, outbox_path, filter_spec=filter_spec)
        if not isinstance(endpoint, str) or not endpoint.startswith(("http://", "https://")):
            raise ValueError(
                f"tn.firehose[{name}]: endpoint must be http(s) URL, got {endpoint!r}"
            )
        if not _PROJECT_ID_RE.match(project_id):
            raise ValueError(
                f"tn.firehose[{name}]: project_id must be a UUID v4 string, got {project_id!r}"
            )
        if key_id is not None and not _KEY_ID_RE.match(key_id):
            raise ValueError(
                f"tn.firehose[{name}]: key_id {key_id!r} not in expected shape"
            )

        self._endpoint = endpoint.rstrip("/")
        self._project_id = project_id
        self._key_id = key_id

        # Phase A stub BEK. Phase B swaps for keystore lookup by key_id.
        self._bek = _stub_bek(project_id, key_id)
        self._aes = AESGCM(self._bek)

        # WebSocket connection lazy-opened on first _publish call. The
        # AsyncHandler worker thread serializes _publish calls so we
        # don't need locking around the connection in the steady state,
        # but use a lock for the open/close transitions which can race
        # with _final_flush.
        self._ws_lock = threading.Lock()
        self._ws: Any = None  # WebSocketApp / WebSocket instance

    # ------------------------------------------------------------------
    # WebSocket lifecycle
    # ------------------------------------------------------------------

    def _ws_url(self) -> str:
        """Convert the http(s) endpoint into a ws(s) URL for the
        per-project firehose stream.

        ``https://vault.tn-proto.org`` + project ``00000000-...`` becomes
        ``wss://vault.tn-proto.org/firehose/00000000-...``. In Phase A
        the path matches what the prototype Worker exposes; production
        will route via the vault, which proxies upstream to the Worker.
        """
        u = urlparse(self._endpoint)
        scheme = "wss" if u.scheme == "https" else "ws"
        path = (u.path.rstrip("/") + f"/firehose/{self._project_id}")
        return urlunparse((scheme, u.netloc, path, "", "", ""))

    def _ensure_connected(self) -> Any:
        """Open the WS lazily and cache it. Returns the live connection.

        Raises ImportError if the optional ``[firehose]`` extra is not
        installed, and any underlying network error if connect fails.
        Network failure propagates to the outbox worker for retry.
        """
        if self._ws is not None:
            return self._ws
        with self._ws_lock:
            if self._ws is not None:
                return self._ws
            try:
                import websocket  # type: ignore[import-not-found]
            except ImportError as e:
                raise ImportError(
                    "tn.firehose requires the `websocket-client` package. "
                    "Install via `pip install 'tn-protocol[firehose]'`."
                ) from e
            ws = websocket.create_connection(
                self._ws_url(),
                timeout=10.0,
                # Production sets a Bearer header from vault-issued JWT;
                # Phase A's scaffold connects unauthenticated against the
                # prototype Worker, which accepts the connection.
            )
            self._ws = ws
            _log.info("[%s] firehose WS connected to %s", self.name, self._ws_url())
            return ws

    def _drop_connection(self) -> None:
        """Close + forget the WS so the next _publish reopens."""
        with self._ws_lock:
            ws = self._ws
            self._ws = None
        if ws is not None:
            try:
                ws.close()
            except Exception:  # noqa: BLE001 — best-effort
                pass

    # ------------------------------------------------------------------
    # AsyncHandler contract
    # ------------------------------------------------------------------

    def _publish(self, envelope: dict[str, Any], raw_line: bytes) -> None:
        """Encrypt the frame under the project BEK and send to the vault.

        Wire format per frame: ``nonce(12) || ciphertext_with_tag(N+16)``.
        AAD is the manifest header tuple ``(project_id, key_id, event_type)``
        so the vault and CF Worker can't substitute frames between
        contexts without detection — even though they can't decrypt the
        body.

        Raises on any network or crypto error so the AsyncHandler outbox
        holds the frame for retry with exponential backoff. The next
        retry reopens the WS as needed.
        """
        event_type = str(envelope.get("event_type") or "")
        aad = f"{self._project_id}|{self._key_id or 'stub'}|{event_type}".encode("utf-8")
        nonce = os.urandom(12)
        try:
            ciphertext = self._aes.encrypt(nonce, raw_line, aad)
        except Exception:
            # AESGCM.encrypt shouldn't fail on valid inputs, but if it
            # does the frame is malformed; raise so the outbox can
            # decide what to do (in practice this is a bug, not a retry).
            raise
        wire = nonce + ciphertext
        try:
            ws = self._ensure_connected()
            ws.send_binary(wire)
        except Exception:
            # Drop the connection so the next attempt reopens. Re-raise
            # so the outbox worker retries with backoff.
            self._drop_connection()
            raise

    def _final_flush(self) -> None:
        """Close the WS on handler shutdown. Best-effort."""
        self._drop_connection()
        _log.debug("[%s] firehose WS closed", self.name)


__all__ = ["TnFirehoseHandler"]

File: ts-sdk/src/handlers/firehose.ts
// TN firehose handler: encrypted log streaming to the TN vault.
//
// Port of python/tn/handlers/firehose.py. Streams each accepted envelope as
// an AES-256-GCM-encrypted binary frame over a WebSocket to the vault's
// per-project firehose stream (`wss://<vault>/firehose/<project_id>`), which
// forwards to the CF firehose Worker. Zero-knowledge: the vault/Worker only
// ever see ciphertext.
//
// Scaffold status (mirrors Python): the BEK is a deterministic *test stub*
// (`_stub_bek`, "do-not-use-in-prod") shared with the Python handler so
// frames are mutually decryptable; real keystore-BEK lookup by `key_id`
// lands later. The frame format is final and byte-identical to
// Python: `nonce(12) || AES-256-GCM(raw_line, aad)`, aad =
// `"<project_id>|<key_id|'stub'>|<event_type>"`.
//
// Like the TS `vault.push` handler (and unlike Python's AsyncHandler), this
// follows the lighter TS pattern: no DurableOutbox/worker. Sends are
// best-effort fire-and-forget on the caller's emit; a dropped connection is
// reopened on the next frame. Durable retry is a documented TS-vs-Python
// divergence tracked with the rest of the async-handler subsystem.

import { createCipheriv, createHash, randomBytes } from "node:crypto";
import { Buffer } from "node:buffer";

import { AsyncTNHandler, type FilterSpec } from "./base.js";

const _PROJECT_ID_RE = /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/;
const _KEY_ID_RE = /^fhk_[A-Za-z2-7]+$/;

/**
 * Deterministic test-stub BEK. Mirrors
 * python/tn/handlers/firehose.py::_stub_bek byte-for-byte so a TS-encrypted
 * frame decrypts under the Python handler's key and vice versa. NOT a real
 * key - the real path swaps this for a keystore lookup by `key_id`.
 *
 * @public
 */
export function stubBek(projectId: string, keyId: string | null): Uint8Array {
  const seed = `${projectId}:${keyId ?? "stub-default"}`;
  return new Uint8Array(
    createHash("sha256")
      .update(Buffer.concat([Buffer.from("phase-a-stub-bek-do-not-use-in-prod:"), Buffer.from(seed)]))
      .digest(),
  );
}

/** `https://host` + project -> `wss://host/firehose/<project>`. Mirrors Python `_ws_url`. */
export function firehoseWsUrl(endpoint: string, projectId: string): string {
  const u = new URL(endpoint);
  const scheme = u.protocol === "https:" ? "wss:" : "ws:";
  const basePath = u.pathname.replace(/\/+$/, "");
  return `${scheme}//${u.host}${basePath}/firehose/${projectId}`;
}

/**
 * Build one firehose wire frame: `nonce(12) || AES-256-GCM(rawLine, aad)`.
 * AAD = `"<projectId>|<keyId|'stub'>|<eventType>"`. Byte-identical to
 * Python's `_publish`. `nonce` is injectable for deterministic tests; omit
 * for a fresh random nonce.
 *
 * @public
 */
export function encryptFirehoseFrame(
  bek: Uint8Array,
  projectId: string,
  keyId: string | null,
  eventType: string,
  rawLine: Uint8Array,
  nonce?: Uint8Array,
): Uint8Array {
  if (bek.length !== 32) throw new Error(`firehose frame: bek must be 32 bytes, got ${bek.length}`);
  const iv = nonce ?? new Uint8Array(randomBytes(12));
  const aad = Buffer.from(`${projectId}|${keyId ?? "stub"}|${eventType}`, "utf-8");
  const cipher = createCipheriv("aes-256-gcm", Buffer.from(bek), Buffer.from(iv));
  cipher.setAAD(aad);
  const ct = Buffer.concat([cipher.update(Buffer.from(rawLine)), cipher.final()]);
  const tag = cipher.getAuthTag();
  // Wire: nonce || ciphertext || tag (AES-GCM tag appended, matching Python's
  // AESGCM.encrypt which returns ciphertext+tag).
  return Uint8Array.from(Buffer.concat([Buffer.from(iv), ct, tag]));
}

export interface FirehoseHandlerOptions {
  endpoint: string;
  projectId: string;
  keyId?: string | null;
  filter?: FilterSpec;
  /** Durable-outbox directory. Frames are enqueued here and delivered with
   *  at-least-once retry by the worker. Required for durability. */
  outboxDir: string;
  maxRetries?: number;
  backoffInitialMs?: number;
  backoffMaxMs?: number;
}

/** Minimal structural type for the WebSocket we send binary frames on. */
interface WsLike {
  send(data: Uint8Array): void;
  close(): void;
  readyState: number;
  addEventListener?(type: string, cb: (...a: unknown[]) => void): void;
  on?(type: string, cb: (...a: unknown[]) => void): void;
}

/**
 * Streaming firehose handler. Mirrors python/tn/handlers/firehose.py.
 */
export class TnFirehoseHandler extends AsyncTNHandler {
  private readonly _endpoint: string;
  private readonly _projectId: string;
  private readonly _keyId: string | null;
  private readonly _bek: Uint8Array;
  private _ws: WsLike | null = null;
  private _connecting: Promise<WsLike> | null = null;

  constructor(name: string, opts: FirehoseHandlerOptions) {
    // Validate before super() (no `this` access — pure arg checks).
    if (typeof opts.endpoint !== "string" || !/^https?:\/\//.test(opts.endpoint)) {
      throw new Error(`tn.firehose[${name}]: endpoint must be an http(s) URL, got ${String(opts.endpoint)}`);
    }
    if (!_PROJECT_ID_RE.test(opts.projectId)) {
      throw new Error(`tn.firehose[${name}]: project_id must be a UUID string, got ${opts.projectId}`);
    }
    if (opts.keyId != null && !_KEY_ID_RE.test(opts.keyId)) {
      throw new Error(`tn.firehose[${name}]: key_id ${opts.keyId} not in expected shape (fhk_<base32>)`);
    }
    const asyncOpts: ConstructorParameters<typeof AsyncTNHandler>[1] = { outboxDir: opts.outboxDir };
    if (opts.filter !== undefined) asyncOpts.filter = opts.filter;
    if (opts.maxRetries !== undefined) asyncOpts.maxRetries = opts.maxRetries;
    if (opts.backoffInitialMs !== undefined) asyncOpts.backoffInitialMs = opts.backoffInitialMs;
    if (opts.backoffMaxMs !== undefined) asyncOpts.backoffMaxMs = opts.backoffMaxMs;
    super(name, asyncOpts);
    this._endpoint = opts.endpoint.replace(/\/+$/, "");
    this._projectId = opts.projectId;
    this._keyId = opts.keyId ?? null;
    this._bek = stubBek(this._projectId, this._keyId);
  }

  override resolved_address(): string | null {
    return firehoseWsUrl(this._endpoint, this._projectId);
  }

  /**
   * Encrypt the frame and send it over the WS. Throws on any
   * connect/send failure so the outbox worker retries with backoff (the
   * frame stays durably queued until a send succeeds).
   */
  protected override async publish(
    envelope: Record<string, unknown>,
    rawLine: string,
  ): Promise<void> {
    const eventType = String(envelope["event_type"] ?? "");
    const frame = encryptFirehoseFrame(
      this._bek,
      this._projectId,
      this._keyId,
      eventType,
      new TextEncoder().encode(rawLine),
    );
    let ws: WsLike;
    try {
      ws = await this._ensureConnected();
      ws.send(frame);
    } catch (e) {
      // Drop the connection so the retry reopens, then re-throw so the
      // outbox holds the frame for redelivery.
      this._ws = null;
      throw e instanceof Error ? e : new Error(String(e));
    }
  }

  protected override finalFlush(): void {
    this._closeWs();
  }

  private _closeWs(): void {
    const ws = this._ws;
    this._ws = null;
    if (ws) {
      try {
        ws.close();
      } catch {
        /* best-effort */
      }
    }
  }

  private async _ensureConnected(): Promise<WsLike> {
    if (this._ws && this._ws.readyState === 1) return this._ws;
    if (this._connecting) return this._connecting;
    this._connecting = this._openWs();
    try {
      this._ws = await this._connecting;
      return this._ws;
    } finally {
      this._connecting = null;
    }
  }

  private async _openWs(): Promise<WsLike> {
    const url = firehoseWsUrl(this._endpoint, this._projectId);
    // Prefer the global WebSocket (Node 22+); fall back to the optional `ws`
    // package (mirrors Python's optional `websocket-client` extra).
    const Ctor =
      (globalThis as { WebSocket?: new (u: string) => WsLike }).WebSocket ??
      (await _loadWsPackage());
    const ws = new Ctor(url) as WsLike & { binaryType?: string };
    if ("binaryType" in ws) ws.binaryType = "arraybuffer";
    await new Promise<void>((resolve, reject) => {
      const onOpen = () => resolve();
      const onErr = (e: unknown) => reject(e instanceof Error ? e : new Error("firehose WS error"));
      if (ws.addEventListener) {
        ws.addEventListener("open", onOpen);
        ws.addEventListener("error", onErr);
      } else if (ws.on) {
        ws.on("open", onOpen);
        ws.on("error", onErr);
      } else {
        resolve();
      }
    });
    return ws;
  }
}

async function _loadWsPackage(): Promise<new (u: string) => WsLike> {
  try {
    // Non-literal specifier so tsc doesn't try to resolve the optional `ws`
    // package at build time (it's not a declared dependency — mirrors
    // Python's optional `websocket-client` extra).
    const wsModule = "ws";
    const mod = (await import(wsModule)) as unknown as { default: new (u: string) => WsLike };
    return mod.default;
  } catch (e) {
    throw new Error(
      "tn.firehose requires a WebSocket implementation. Either run on Node 22+ " +
        "(global WebSocket) or install the optional `ws` package.",
      { cause: e },
    );
  }
}
